### ISSUE TITLE
MSVC and GCC Linking Errors to missing extern "Rust" function symbols

### ISSUE THREAD

**Issue by david-cattermole** [dtolnay/cxx#434]
Hello David,

Firstly, thank you for your amazing contributions to the world of Rust!

Secondly, I have begun testing your crate and have experienced some problems with the adoption on a small, but real-world example library I am starting on. I am very eager to start using the crate!

I have made an example project on GitHub to potentially help others, but also to explain the issues that I'm having.
https://github.com/david-cattermole/cxx-demo-example

### Linking Errors
When building and linking the Rust library into a shared library (.so / .dll), I am getting linking errors on both Windows MSVC and CentOS Linux with GCC.

GCC 6.3.1:
```
CMakeFiles/mmscenegraph_tests.dir/tests/test_b.cpp.o: In function `test_b()':
test_b.cpp:(.text+0x65): undefined reference to `rust::cxxbridge05::Str::Str(std::string const&)'
CMakeFiles/mmscenegraph_tests.dir/src/lib.cpp.o: In function `mmscenegraph::make_demo(rust::cxxbridge05::Str)':
lib.cpp:(.text+0x14b): undefined reference to `rust::cxxbridge05::Str::operator std::string() const'
CMakeFiles/mmscenegraph_tests.dir/src/_cxxbridge.cpp.o: In function `mmscenegraph$cxxbridge05$make_demo':
_cxxbridge.cpp:(.text+0x10): undefined reference to `rust::cxxbridge05::Str::Str()'
CMakeFiles/mmscenegraph_tests.dir/src/_cxxbridge.cpp.o: In function `mmscenegraph$cxxbridge05$do_thing':
_cxxbridge.cpp:(.text+0x9c): undefined reference to `cxxbridge05$box$mmscenegraph$ThingR$drop'
CMakeFiles/mmscenegraph_tests.dir/src/_cxxbridge.cpp.o: In function `mmscenegraph::print_r(mmscenegraph::ThingR const&)':
_cxxbridge.cpp:(.text+0xb1): undefined reference to `mmscenegraph$cxxbridge05$print_r'
CMakeFiles/mmscenegraph_tests.dir/src/_cxxbridge.cpp.o: In function `rust::cxxbridge05::Box<mmscenegraph::ThingR>::uninit()':
_cxxbridge.cpp:(.text+0x131): undefined reference to `cxxbridge05$box$mmscenegraph$ThingR$uninit'
CMakeFiles/mmscenegraph_tests.dir/src/_cxxbridge.cpp.o: In function `rust::cxxbridge05::Box<mmscenegraph::ThingR>::drop()':
_cxxbridge.cpp:(.text+0x141): undefined reference to `cxxbridge05$box$mmscenegraph$ThingR$drop'
```

Visual Studio 2015:
```
CMakeFiles/mmscenegraph_tests.dir/tests/test_b.cpp.o: In function `test_b()':
test_b.cpp:(.text+0x65): undefined reference to `rust::cxxbridge05::Str::Str(std::string const&)'
CMakeFiles/mmscenegraph_tests.dir/src/lib.cpp.o: In function `mmscenegraph::make_demo(rust::cxxbridge05::Str)':
lib.cpp:(.text+0x14b): undefined reference to `rust::cxxbridge05::Str::operator std::string() const'
CMakeFiles/mmscenegraph_tests.dir/src/_cxxbridge.cpp.o: In function `mmscenegraph$cxxbridge05$make_demo':
_cxxbridge.cpp:(.text+0x10): undefined reference to `rust::cxxbridge05::Str::Str()'
CMakeFiles/mmscenegraph_tests.dir/src/_cxxbridge.cpp.o: In function `mmscenegraph$cxxbridge05$do_thing':
_cxxbridge.cpp:(.text+0x9c): undefined reference to `cxxbridge05$box$mmscenegraph$ThingR$drop'
CMakeFiles/mmscenegraph_tests.dir/src/_cxxbridge.cpp.o: In function `mmscenegraph::print_r(mmscenegraph::ThingR const&)':
_cxxbridge.cpp:(.text+0xb1): undefined reference to `mmscenegraph$cxxbridge05$print_r'
CMakeFiles/mmscenegraph_tests.dir/src/_cxxbridge.cpp.o: In function `rust::cxxbridge05::Box<mmscenegraph::ThingR>::uninit()':
_cxxbridge.cpp:(.text+0x131): undefined reference to `cxxbridge05$box$mmscenegraph$ThingR$uninit'
CMakeFiles/mmscenegraph_tests.dir/src/_cxxbridge.cpp.o: In function `rust::cxxbridge05::Box<mmscenegraph::ThingR>::drop()':
_cxxbridge.cpp:(.text+0x141): undefined reference to `cxxbridge05$box$mmscenegraph$ThingR$drop'
```

I am using CMake for the C++ side, and Cargo for the Rust side, with a Bash and Windows Batch script gluing them together.

My approach is to generate the headers using cxx, and cbindgen, then compile the Rust crate as a "staticlib" and link the Rust static library with a C++ shared library, combining the Rust symbols into the C++ shared library, with a single C++ interface.

I have noticed that most (but not all) of the problem functions are defined in the `extern "Rust"` block here:
https://github.com/david-cattermole/cxx-demo-example/blob/main/src/cxxbridge.rs#L22

I have also noticed that the generated C++ functions do not implement those functions and that is the true cause for the missing symbols at linking stage. For example `mmscenegraph$cxxbridge05$print_r`:
https://github.com/david-cattermole/cxx-demo-example/blob/main/src/_cxxbridge.cpp#L243

You can see the full output of the `build_linux.bash` and `build_windows.bat` scripts:
[output_gcc_6_3_1.log](https://github.com/dtolnay/cxx/files/5504604/output_gcc_6_3_1.log) or [here on GitHub](https://github.com/david-cattermole/cxx-demo-example/blob/main/output_gcc_6_3_1.log)
[output_msvc_19_0_24215_1.log](https://github.com/dtolnay/cxx/files/5504605/output_msvc_19_0_24215_1.log) or [here on GitHub](https://github.com/david-cattermole/cxx-demo-example/blob/main/output_msvc_19_0_24215_1.log)

I have also attempted to use "cxx v0.4.7", but I seem to get the same problem.

Am I missing something? Is my use case not supported? Is this a bug in cxx? I am I misusing cxx?

### Example Project Goals
For context, the goal for the example project is to write a cross platform Rust and C++ library, with the main client being an existing C++ application, taking advantage of Rust whenever possible.
I also intend to call other C++ library dependencies, integrating well established C++ libraries that are dominate in my industry (Visual Effects).
Therefore I will use cxx/cxxbridge to create the C++ bindings, and then fallback on cbindgen whenever I hit a missing feature.

Thank you,
David Cattermole

**Comment by dtolnay** [comment]
From the GCC output the symbols it doesn't have are:

rust::cxxbridge05::Str::Str()
rust::cxxbridge05::Str::Str(std::string const&)
rust::cxxbridge05::Str::operator std::string() const
cxxbridge05$box$mmscenegraph$ThingR$drop
cxxbridge05$box$mmscenegraph$ThingR$uninit
mmscenegraph$cxxbridge05$print_r

The first 3 are all defined in https://github.com/dtolnay/cxx/blob/0.5.5/src/cxx.cc. I haven't looked at your repro but maybe you haven't built or linked that?

The remaining 3 are defined in Rust by the Rust library containing the #\[cxx::bridge\] macro invocation.

FYI there is an example of a working CMake setup here: https://github.com/XiangpengHao/cxx-cmake-example.

**Comment by david-cattermole** [comment]
Thanks for the quick reply.

I have copied cxx.cc and included the file in my shared object, unfortunately I now have 70 missing symbols, but since including cxx.cc the 3 symbols are now not erroring:
```
rust::cxxbridge05::Str::Str()
rust::cxxbridge05::Str::Str(std::string const&)
rust::cxxbridge05::Str::operator std::string() const
```

I have tested on MSVC only, not on Linux with GCC.
MSVC output:
```
[  9%] Linking CXX shared library libmmscenegraph.dll
LINK: command "C:\PROGRA~2\MICROS~2.0\VC\bin\X86_AM~1\link.exe /nologo @CMakeFiles\libmmscenegraph.dir\objects1.rsp /out:libmmscenegraph.dll /implib:libmmscenegraph.lib /pdb:C:\Users\catte\dev\cxx-demo-example\build\libmmscenegraph.pdb /dll /version:0.0 /machine:x64 /INCREMENTAL:NO ..\target\release\mmscenegraph.lib ws2_32.lib userenv.lib advapi32.lib shell32.lib msvcrt.lib kernel32.lib user32.lib gdi32.lib winspool.lib shell32.lib ole32.lib oleaut32.lib uuid.lib comdlg32.lib advapi32.lib /MANIFEST /MANIFESTFILE:libmmscenegraph.dll.manifest" failed (exit code 1120) with the following output:
   Creating library libmmscenegraph.lib and object libmmscenegraph.exp
_cxxbridge.cpp.obj : error LNK2019: unresolved external symbol mmscenegraph$cxxbridge05$print_r referenced in function "void __cdecl mmscenegraph::print_r(struct mmscenegraph::ThingR const &)" (?print_r@mmscenegraph@@YAXAEBUThingR@1@@Z)
_cxxbridge.cpp.obj : error LNK2019: unresolved external symbol cxxbridge05$box$mmscenegraph$ThingR$uninit referenced in function "private: void __cdecl rust::cxxbridge05::Box<struct mmscenegraph::ThingR>::uninit(void)" (?uninit@?$Box@UThingR@mmscenegraph@@@cxxbridge05@rust@@AEAAXXZ)
_cxxbridge.cpp.obj : error LNK2019: unresolved external symbol cxxbridge05$box$mmscenegraph$ThingR$drop referenced in function "public: __cdecl rust::cxxbridge05::Box<struct mmscenegraph::ThingR>::~Box<struct mmscenegraph::ThingR>(void)" (??1?$Box@UThingR@mmscenegraph@@@cxxbridge05@rust@@QEAA@XZ)
cxx.cc.obj : error LNK2019: unresolved external symbol cxxbridge05$string$new referenced in function "public: __cdecl rust::cxxbridge05::String::String(class rust::cxxbridge05::String &&)" (??0String@cxxbridge05@rust@@QEAA@$$QEAV012@@Z)
cxx.cc.obj : error LNK2019: unresolved external symbol cxxbridge05$string$clone referenced in function "public: __cdecl rust::cxxbridge05::String::String(class rust::cxxbridge05::String const &)" (??0String@cxxbridge05@rust@@QEAA@AEBV012@@Z)
cxx.cc.obj : error LNK2019: unresolved external symbol cxxbridge05$string$from referenced in function "public: __cdecl rust::cxxbridge05::String::String(class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > const &)" (??0String@cxxbridge05@rust@@QEAA@AEBV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@@Z)
cxx.cc.obj : error LNK2019: unresolved external symbol cxxbridge05$string$drop referenced in function "public: __cdecl rust::cxxbridge05::String::~String(void)" (??1String@cxxbridge05@rust@@QEAA@XZ)
cxx.cc.obj : error LNK2019: unresolved external symbol cxxbridge05$string$ptr referenced in function "class std::basic_ostream<char,struct std::char_traits<char> > & __cdecl rust::cxxbridge05::operator<<(class std::basic_ostream<char,struct std::char_traits<char> > &,class rust::cxxbridge05::String const &)" (??6cxxbridge05@rust@@YAAEAV?$basic_ostream@DU?$char_traits@D@std@@@std@@AEAV23@AEBVString@01@@Z)
cxx.cc.obj : error LNK2019: unresolved external symbol cxxbridge05$string$len referenced in function "class std::basic_ostream<char,struct std::char_traits<char> > & __cdecl rust::cxxbridge05::operator<<(class std::basic_ostream<char,struct std::char_traits<char> > &,class rust::cxxbridge05::String const &)" (??6cxxbridge05@rust@@YAAEAV?$basic_ostream@DU?$char_traits@D@std@@@std@@AEAV23@AEBVString@01@@Z)
cxx.cc.obj : error LNK2019: unresolved external symbol cxxbridge05$str$valid referenced in function "public: __cdecl rust::cxxbridge05::Str::Str(class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > const &)" (??0Str@cxxbridge05@rust@@QEAA@AEBV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@@Z)
cxx.cc.obj : error LNK2019: unresolved external symbol cxxbridge05$rust_vec$u8$new referenced in function "public: __cdecl rust::cxxbridge05::Vec<unsigned char>::Vec<unsigned char>(void)" (??0?$Vec@E@cxxbridge05@rust@@QEAA@XZ)
cxx.cc.obj : error LNK2019: unresolved external symbol cxxbridge05$rust_vec$u8$drop referenced in function "private: void __cdecl rust::cxxbridge05::Vec<unsigned char>::drop(void)" (?drop@?$Vec@E@cxxbridge05@rust@@AEAAXXZ)
cxx.cc.obj : error LNK2019: unresolved external symbol cxxbridge05$rust_vec$u8$len referenced in function "public: unsigned __int64 __cdecl rust::cxxbridge05::Vec<unsigned char>::size(void)const " (?size@?$Vec@E@cxxbridge05@rust@@QEBA_KXZ)
cxx.cc.obj : error LNK2019: unresolved external symbol cxxbridge05$rust_vec$u8$data referenced in function "public: unsigned char const * __cdecl rust::cxxbridge05::Vec<unsigned char>::data(void)const " (?data@?$Vec@E@cxxbridge05@rust@@QEBAPEBEXZ)
cxx.cc.obj : error LNK2019: unresolved external symbol cxxbridge05$rust_vec$u8$stride referenced in function "private: static unsigned __int64 __cdecl rust::cxxbridge05::Vec<unsigned char>::stride(void)" (?stride@?$Vec@E@cxxbridge05@rust@@CA_KXZ)
cxx.cc.obj : error LNK2019: unresolved external symbol cxxbridge05$rust_vec$u16$new referenced in function "public: __cdecl rust::cxxbridge05::Vec<unsigned short>::Vec<unsigned short>(void)" (??0?$Vec@G@cxxbridge05@rust@@QEAA@XZ)
cxx.cc.obj : error LNK2019: unresolved external symbol cxxbridge05$rust_vec$u16$drop referenced in function "private: void __cdecl rust::cxxbridge05::Vec<unsigned short>::drop(void)" (?drop@?$Vec@G@cxxbridge05@rust@@AEAAXXZ)
cxx.cc.obj : error LNK2019: unresolved external symbol cxxbridge05$rust_vec$u16$len referenced in function "public: unsigned __int64 __cdecl rust::cxxbridge05::Vec<unsigned short>::size(void)const " (?size@?$Vec@G@cxxbridge05@rust@@QEBA_KXZ)
cxx.cc.obj : error LNK2019: unresolved external symbol cxxbridge05$rust_vec$u16$data referenced in function "public: unsigned short const * __cdecl rust::cxxbridge05::Vec<unsigned short>::data(void)const " (?data@?$Vec@G@cxxbridge05@rust@@QEBAPEBGXZ)
cxx.cc.obj : error LNK2019: unresolved external symbol cxxbridge05$rust_vec$u16$stride referenced in function "private: static unsigned __int64 __cdecl rust::cxxbridge05::Vec<unsigned short>::stride(void)" (?stride@?$Vec@G@cxxbridge05@rust@@CA_KXZ)
cxx.cc.obj : error LNK2019: unresolved external symbol cxxbridge05$rust_vec$u32$new referenced in function "public: __cdecl rust::cxxbridge05::Vec<unsigned int>::Vec<unsigned int>(void)" (??0?$Vec@I@cxxbridge05@rust@@QEAA@XZ)
cxx.cc.obj : error LNK2019: unresolved external symbol cxxbridge05$rust_vec$u32$drop referenced in function "private: void __cdecl rust::cxxbridge05::Vec<unsigned int>::drop(void)" (?drop@?$Vec@I@cxxbridge05@rust@@AEAAXXZ)
cxx.cc.obj : error LNK2019: unresolved external symbol cxxbridge05$rust_vec$u32$len referenced in function "public: unsigned __int64 __cdecl rust::cxxbridge05::Vec<unsigned int>::size(void)const " (?size@?$Vec@I@cxxbridge05@rust@@QEBA_KXZ)
cxx.cc.obj : error LNK2019: unresolved external symbol cxxbridge05$rust_vec$u32$data referenced in function "public: unsigned int const * __cdecl rust::cxxbridge05::Vec<unsigned int>::data(void)const " (?data@?$Vec@I@cxxbridge05@rust@@QEBAPEBIXZ)
cxx.cc.obj : error LNK2019: unresolved external symbol cxxbridge05$rust_vec$u32$stride referenced in function "private: static unsigned __int64 __cdecl rust::cxxbridge05::Vec<unsigned int>::stride(void)" (?stride@?$Vec@I@cxxbridge05@rust@@CA_KXZ)
cxx.cc.obj : error LNK2019: unresolved external symbol cxxbridge05$rust_vec$u64$new referenced in function "public: __cdecl rust::cxxbridge05::Vec<unsigned __int64>::Vec<unsigned __int64>(void)" (??0?$Vec@_K@cxxbridge05@rust@@QEAA@XZ)
cxx.cc.obj : error LNK2019: unresolved external symbol cxxbridge05$rust_vec$u64$drop referenced in function "private: void __cdecl rust::cxxbridge05::Vec<unsigned __int64>::drop(void)" (?drop@?$Vec@_K@cxxbridge05@rust@@AEAAXXZ)
cxx.cc.obj : error LNK2019: unresolved external symbol cxxbridge05$rust_vec$u64$len referenced in function "public: unsigned __int64 __cdecl rust::cxxbridge05::Vec<unsigned __int64>::size(void)const " (?size@?$Vec@_K@cxxbridge05@rust@@QEBA_KXZ)
cxx.cc.obj : error LNK2019: unresolved external symbol cxxbridge05$rust_vec$u64$data referenced in function "public: unsigned __int64 const * __cdecl rust::cxxbridge05::Vec<unsigned __int64>::data(void)const " (?data@?$Vec@_K@cxxbridge05@rust@@QEBAPEB_KXZ)
cxx.cc.obj : error LNK2019: unresolved external symbol cxxbridge05$rust_vec$u64$stride referenced in function "private: static unsigned __int64 __cdecl rust::cxxbridge05::Vec<unsigned __int64>::stride(void)" (?stride@?$Vec@_K@cxxbridge05@rust@@CA_KXZ)
cxx.cc.obj : error LNK2019: unresolved external symbol cxxbridge05$rust_vec$i8$new referenced in function "public: __cdecl rust::cxxbridge05::Vec<signed char>::Vec<signed char>(void)" (??0?$Vec@C@cxxbridge05@rust@@QEAA@XZ)
cxx.cc.obj : error LNK2019: unresolved external symbol cxxbridge05$rust_vec$i8$drop referenced in function "private: void __cdecl rust::cxxbridge05::Vec<signed char>::drop(void)" (?drop@?$Vec@C@cxxbridge05@rust@@AEAAXXZ)
cxx.cc.obj : error LNK2019: unresolved external symbol cxxbridge05$rust_vec$i8$len referenced in function "public: unsigned __int64 __cdecl rust::cxxbridge05::Vec<signed char>::size(void)const " (?size@?$Vec@C@cxxbridge05@rust@@QEBA_KXZ)
cxx.cc.obj : error LNK2019: unresolved external symbol cxxbridge05$rust_vec$i8$data referenced in function "public: signed char const * __cdecl rust::cxxbridge05::Vec<signed char>::data(void)const " (?data@?$Vec@C@cxxbridge05@rust@@QEBAPEBCXZ)
cxx.cc.obj : error LNK2019: unresolved external symbol cxxbridge05$rust_vec$i8$stride referenced in function "private: static unsigned __int64 __cdecl rust::cxxbridge05::Vec<signed char>::stride(void)" (?stride@?$Vec@C@cxxbridge05@rust@@CA_KXZ)
cxx.cc.obj : error LNK2019: unresolved external symbol cxxbridge05$rust_vec$i16$new referenced in function "public: __cdecl rust::cxxbridge05::Vec<short>::Vec<short>(void)" (??0?$Vec@F@cxxbridge05@rust@@QEAA@XZ)
cxx.cc.obj : error LNK2019: unresolved external symbol cxxbridge05$rust_vec$i16$drop referenced in function "private: void __cdecl rust::cxxbridge05::Vec<short>::drop(void)" (?drop@?$Vec@F@cxxbridge05@rust@@AEAAXXZ)
cxx.cc.obj : error LNK2019: unresolved external symbol cxxbridge05$rust_vec$i16$len referenced in function "public: unsigned __int64 __cdecl rust::cxxbridge05::Vec<short>::size(void)const " (?size@?$Vec@F@cxxbridge05@rust@@QEBA_KXZ)
cxx.cc.obj : error LNK2019: unresolved external symbol cxxbridge05$rust_vec$i16$data referenced in function "public: short const * __cdecl rust::cxxbridge05::Vec<short>::data(void)const " (?data@?$Vec@F@cxxbridge05@rust@@QEBAPEBFXZ)
cxx.cc.obj : error LNK2019: unresolved external symbol cxxbridge05$rust_vec$i16$stride referenced in function "private: static unsigned __int64 __cdecl rust::cxxbridge05::Vec<short>::stride(void)" (?stride@?$Vec@F@cxxbridge05@rust@@CA_KXZ)
cxx.cc.obj : error LNK2019: unresolved external symbol cxxbridge05$rust_vec$i32$new referenced in function "public: __cdecl rust::cxxbridge05::Vec<int>::Vec<int>(void)" (??0?$Vec@H@cxxbridge05@rust@@QEAA@XZ)
cxx.cc.obj : error LNK2019: unresolved external symbol cxxbridge05$rust_vec$i32$drop referenced in function "private: void __cdecl rust::cxxbridge05::Vec<int>::drop(void)" (?drop@?$Vec@H@cxxbridge05@rust@@AEAAXXZ)
cxx.cc.obj : error LNK2019: unresolved external symbol cxxbridge05$rust_vec$i32$len referenced in function "public: unsigned __int64 __cdecl rust::cxxbridge05::Vec<int>::size(void)const " (?size@?$Vec@H@cxxbridge05@rust@@QEBA_KXZ)
cxx.cc.obj : error LNK2019: unresolved external symbol cxxbridge05$rust_vec$i32$data referenced in function "public: int const * __cdecl rust::cxxbridge05::Vec<int>::data(void)const " (?data@?$Vec@H@cxxbridge05@rust@@QEBAPEBHXZ)
cxx.cc.obj : error LNK2019: unresolved external symbol cxxbridge05$rust_vec$i32$stride referenced in function "private: static unsigned __int64 __cdecl rust::cxxbridge05::Vec<int>::stride(void)" (?stride@?$Vec@H@cxxbridge05@rust@@CA_KXZ)
cxx.cc.obj : error LNK2019: unresolved external symbol cxxbridge05$rust_vec$i64$new referenced in function "public: __cdecl rust::cxxbridge05::Vec<__int64>::Vec<__int64>(void)" (??0?$Vec@_J@cxxbridge05@rust@@QEAA@XZ)
cxx.cc.obj : error LNK2019: unresolved external symbol cxxbridge05$rust_vec$i64$drop referenced in function "private: void __cdecl rust::cxxbridge05::Vec<__int64>::drop(void)" (?drop@?$Vec@_J@cxxbridge05@rust@@AEAAXXZ)
cxx.cc.obj : error LNK2019: unresolved external symbol cxxbridge05$rust_vec$i64$len referenced in function "public: unsigned __int64 __cdecl rust::cxxbridge05::Vec<__int64>::size(void)const " (?size@?$Vec@_J@cxxbridge05@rust@@QEBA_KXZ)
cxx.cc.obj : error LNK2019: unresolved external symbol cxxbridge05$rust_vec$i64$data referenced in function "public: __int64 const * __cdecl rust::cxxbridge05::Vec<__int64>::data(void)const " (?data@?$Vec@_J@cxxbridge05@rust@@QEBAPEB_JXZ)
cxx.cc.obj : error LNK2019: unresolved external symbol cxxbridge05$rust_vec$i64$stride referenced in function "private: static unsigned __int64 __cdecl rust::cxxbridge05::Vec<__int64>::stride(void)" (?stride@?$Vec@_J@cxxbridge05@rust@@CA_KXZ)
cxx.cc.obj : error LNK2019: unresolved external symbol cxxbridge05$rust_vec$f32$new referenced in function "public: __cdecl rust::cxxbridge05::Vec<float>::Vec<float>(void)" (??0?$Vec@M@cxxbridge05@rust@@QEAA@XZ)
cxx.cc.obj : error LNK2019: unresolved external symbol cxxbridge05$rust_vec$f32$drop referenced in function "private: void __cdecl rust::cxxbridge05::Vec<float>::drop(void)" (?drop@?$Vec@M@cxxbridge05@rust@@AEAAXXZ)
cxx.cc.obj : error LNK2019: unresolved external symbol cxxbridge05$rust_vec$f32$len referenced in function "public: unsigned __int64 __cdecl rust::cxxbridge05::Vec<float>::size(void)const " (?size@?$Vec@M@cxxbridge05@rust@@QEBA_KXZ)
cxx.cc.obj : error LNK2019: unresolved external symbol cxxbridge05$rust_vec$f32$data referenced in function "public: float const * __cdecl rust::cxxbridge05::Vec<float>::data(void)const " (?data@?$Vec@M@cxxbridge05@rust@@QEBAPEBMXZ)
cxx.cc.obj : error LNK2019: unresolved external symbol cxxbridge05$rust_vec$f32$stride referenced in function "private: static unsigned __int64 __cdecl rust::cxxbridge05::Vec<float>::stride(void)" (?stride@?$Vec@M@cxxbridge05@rust@@CA_KXZ)
cxx.cc.obj : error LNK2019: unresolved external symbol cxxbridge05$rust_vec$f64$new referenced in function "public: __cdecl rust::cxxbridge05::Vec<double>::Vec<double>(void)" (??0?$Vec@N@cxxbridge05@rust@@QEAA@XZ)
cxx.cc.obj : error LNK2019: unresolved external symbol cxxbridge05$rust_vec$f64$drop referenced in function "private: void __cdecl rust::cxxbridge05::Vec<double>::drop(void)" (?drop@?$Vec@N@cxxbridge05@rust@@AEAAXXZ)
cxx.cc.obj : error LNK2019: unresolved external symbol cxxbridge05$rust_vec$f64$len referenced in function "public: unsigned __int64 __cdecl rust::cxxbridge05::Vec<double>::size(void)const " (?size@?$Vec@N@cxxbridge05@rust@@QEBA_KXZ)
cxx.cc.obj : error LNK2019: unresolved external symbol cxxbridge05$rust_vec$f64$data referenced in function "public: double const * __cdecl rust::cxxbridge05::Vec<double>::data(void)const " (?data@?$Vec@N@cxxbridge05@rust@@QEBAPEBNXZ)
cxx.cc.obj : error LNK2019: unresolved external symbol cxxbridge05$rust_vec$f64$stride referenced in function "private: static unsigned __int64 __cdecl rust::cxxbridge05::Vec<double>::stride(void)" (?stride@?$Vec@N@cxxbridge05@rust@@CA_KXZ)
cxx.cc.obj : error LNK2019: unresolved external symbol cxxbridge05$rust_vec$bool$new referenced in function "public: __cdecl rust::cxxbridge05::Vec<bool>::Vec<bool>(void)" (??0?$Vec@_N@cxxbridge05@rust@@QEAA@XZ)
cxx.cc.obj : error LNK2019: unresolved external symbol cxxbridge05$rust_vec$bool$drop referenced in function "private: void __cdecl rust::cxxbridge05::Vec<bool>::drop(void)" (?drop@?$Vec@_N@cxxbridge05@rust@@AEAAXXZ)
cxx.cc.obj : error LNK2019: unresolved external symbol cxxbridge05$rust_vec$bool$len referenced in function "public: unsigned __int64 __cdecl rust::cxxbridge05::Vec<bool>::size(void)const " (?size@?$Vec@_N@cxxbridge05@rust@@QEBA_KXZ)
cxx.cc.obj : error LNK2019: unresolved external symbol cxxbridge05$rust_vec$bool$data referenced in function "public: bool const * __cdecl rust::cxxbridge05::Vec<bool>::data(void)const " (?data@?$Vec@_N@cxxbridge05@rust@@QEBAPEB_NXZ)
cxx.cc.obj : error LNK2019: unresolved external symbol cxxbridge05$rust_vec$bool$stride referenced in function "private: static unsigned __int64 __cdecl rust::cxxbridge05::Vec<bool>::stride(void)" (?stride@?$Vec@_N@cxxbridge05@rust@@CA_KXZ)
cxx.cc.obj : error LNK2019: unresolved external symbol cxxbridge05$rust_vec$string$new referenced in function "public: __cdecl rust::cxxbridge05::Vec<class rust::cxxbridge05::String>::Vec<class rust::cxxbridge05::String>(void)" (??0?$Vec@VString@cxxbridge05@rust@@@cxxbridge05@rust@@QEAA@XZ)
cxx.cc.obj : error LNK2019: unresolved external symbol cxxbridge05$rust_vec$string$drop referenced in function "private: void __cdecl rust::cxxbridge05::Vec<class rust::cxxbridge05::String>::drop(void)" (?drop@?$Vec@VString@cxxbridge05@rust@@@cxxbridge05@rust@@AEAAXXZ)
cxx.cc.obj : error LNK2019: unresolved external symbol cxxbridge05$rust_vec$string$len referenced in function "public: unsigned __int64 __cdecl rust::cxxbridge05::Vec<class rust::cxxbridge05::String>::size(void)const " (?size@?$Vec@VString@cxxbridge05@rust@@@cxxbridge05@rust@@QEBA_KXZ)
cxx.cc.obj : error LNK2019: unresolved external symbol cxxbridge05$rust_vec$string$data referenced in function "public: class rust::cxxbridge05::String const * __cdecl rust::cxxbridge05::Vec<class rust::cxxbridge05::String>::data(void)const " (?data@?$Vec@VString@cxxbridge05@rust@@@cxxbridge05@rust@@QEBAPEBVString@23@XZ)
cxx.cc.obj : error LNK2019: unresolved external symbol cxxbridge05$rust_vec$string$stride referenced in function "private: static unsigned __int64 __cdecl rust::cxxbridge05::Vec<class rust::cxxbridge05::String>::stride(void)" (?stride@?$Vec@VString@cxxbridge05@rust@@@cxxbridge05@rust@@CA_KXZ)
libmmscenegraph.dll : fatal error LNK1120: 70 unresolved externals
```

The first 3 missing symbols are coming from a file generated with:
```
cxxbridge src/cxxbridge.rs --header --output include\mmscenegraph\_cxxbridge.h
cxxbridge src/cxxbridge.rs --cxx-impl-annotations "__declspec(dllexport)" --output src\_cxxbridge.cpp
```
I would expect `_cxxbridge.cpp` to have these symbols implemented.

P.S. Do I need to use the nightly rust compiler?
My active toolchain is: 
```
stable-x86_64-pc-windows-msvc (default)
rustc 1.47.0 (18bf6b4f0 2020-10-07)
```

Regarding the `cxx.cc` file, I didn't realise I needed the file.
Would it be possible to add it to the `cxxbridge` command to automatically generate the file to avoid the file becoming out of date with the installed cxxbridge-cmd and cxx crate versions?

P.S. Unfortunately https://github.com/XiangpengHao/cxx-cmake-example does not compile for me either, but it seems like a different problem.

**Comment by dtolnay** [comment]
> I would expect `_cxxbridge.cpp` to have these symbols implemented.

No because those symbols are implemented in Rust (per https://github.com/dtolnay/cxx/issues/434#issuecomment-723453231) so they cannot be implemented by a C++ file.

The rest are implemented also in Rust, by the cxx crate.

It seems your CMake setup does not provide Rust artifacts to the linker.

**Comment by dtolnay** [comment]
> P.S. Unfortunately https://github.com/XiangpengHao/cxx-cmake-example does not compile for me either, but it seems like a different problem.

If you can provide some detail about this, @XiangpengHao may be able to help out.

**Comment by david-cattermole** [comment]
Good news, I've got it all working now!

As expected it was user error, not cxx that was the problem.

The reason for my missing symbols was that I didn't have `cxxbridge.rs` as a public module in my `lib.rs` file:
https://github.com/david-cattermole/cxx-demo-example/blob/main/src/lib.rs

I also had problems with duplicate definitions for the `extern "C"` functions since I am also running cbindgen on the project to produce manual C bindings, these were incorrectly being defined twice. I fixed this by adding an "[export] exclude" key into the [cbindgen.toml](https://github.com/david-cattermole/cxx-demo-example/blob/main/cbindgen.toml#L153)

In summary, my biggest problems were:

1. I didn't know I needed to copy `cxx.cc` from the cxx repo and link with it.

2. I was not including the file with `#[cxx::bridge(namespace = ...)]` in my `lib.rs` file, hence it was not compiled and so the symbols were missing.

3. cbindgen generated duplicate definitions for the `extern C` functions that were exposed via `#[cxx::bridge(namespace = ...)]`. I manually excluded them in the `cbindgen.toml` file.

4. Incorrectly assuming the [demo](https://github.com/dtolnay/cxx/tree/0.5.5/demo) was C++11 (#435) - it is a C++14 example.

5. Not knowing what type of Rust `crate_type` I should use and what type C++ library (shared or static) would be best. (for Rust it should be `staticlib` and for C++ it should be `SHARED`, although it is probably possible to also use `STATIC` for the C++ side, I prefer a final .so/.dll library.

Also, as some personal advice to anyone reading this in my same position, I would recommend using Linux and GCC for debugging not MSVC, as it's very easy to see the symbols in Linux with the command:
```
$ nm -CD path/to/lib.so | grep " T "
```

You can see the output for GCC and MSVC here:
[Output under GCC 6.3.1](https://github.com/david-cattermole/cxx-demo-example/blob/main/output_gcc_6_3_1.log)
[Output under MSVC 16 (Visual Studio 2015)](https://github.com/david-cattermole/cxx-demo-example/blob/main/output_msvc_19_0_24215_1.log)

Thanks again @dtolnay!

**Comment by david-cattermole** [comment]
Since this is working and I have documented the example project, I think we can close this.

https://github.com/david-cattermole/cxx-demo-example/blob/main/README.md#project-anatomy

Thanks again.

**Comment by dtolnay** [comment]
That's great! Some notes:

- The fork of cxx.cc should not be necessary. That is already compiled into the cxx crate by Cargo.

- `./build_linux.bash` does not run because it is not executable.

    ```console
    $ ./build_linux.bash
    bash: ./build_linux.bash: Permission denied
    ```

- It would be better if it didn't hardcode CC to /opt/rh/devtoolset-6/root/usr/bin/gcc since that makes the build fail on most systems and seems like the only thing that is centos specific.

**Comment by david-cattermole** [comment]
Thanks David,

I have applied your suggestions, removing references to `cxx.cc` and removing the defaults of the CentOS 7.8 (with a small note)

It seems the reason `cxx.cc` was needed in my example was that I added
```
[target.x86_64-pc-windows-msvc]
rustflags = ["-C", "target-feature=+crt-static"]
```
into the `.cargo/config` file. I did this while trying to fix the symbol linking problem and never deleted it.
Now it's deleted, it works perfectly.
